recipes/dev/generate_v2_distributed.py

```diff
@@ -0,0 +1,259 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+import itertools
+import sys
+import time
+from typing import Any, Dict, List
+
+import torch
+import torch.distributed as dist
+from omegaconf import DictConfig, OmegaConf
+from torch.distributed.tensor.parallel import parallelize_module
+
+from torchtune import config, training, utils
+from torchtune.data import load_image, Message, padded_collate_tiled_images_and_mask
+
+from torchtune.generation import sample
+
+from torchtune.modules.transforms import Transform
+
+
+class SingleTurnYAMLToMessages(Transform):
+    """
+    Converts a single turn conversation in YAML format to a list of messages.
+
+    Expects the YAML to look like:
+        system: You are a helpful AI assistant.
+        user: What is the capital of France?
+
+    or if it includes an image:
+        system: You are a helpful AI assistant.
+        user:
+            image: url or path_to_image
+            text: Describe the image in detail.
+    """
+
+    def __call__(self, prompt: Dict[str, Any]) -> List[Message]:
+        messages = []
+
+        # Iterate through roles and add content
+        for role, content in prompt.items():
+            if content is None:
+                continue
+            elif isinstance(content, str):
+                new_content = [{"type": "text", "content": content}]
+            elif "image" in content.keys():
+                image_loc = content["image"]
+                image = load_image(image_loc)
+                new_content = [
+                    {"type": "image", "content": image},
+                    {"type": "text", "content": content["text"]},
+                ]
+            else:
+                assert (
+                    "text" in content.keys()
+                ), "Multiple entries per role expect at least a text key"
+                new_content = [{"type": "text", "content": content["text"]}]
+            messages.append(Message(role=role, content=new_content))
+
+        # Finally, add an empty assistant message to kick-start generation
+        messages.append(Message(role="assistant", content=""))
+        return messages
+
+
+class InferenceRecipe:
+    """
+    Recipe for generating tokens from a dense Transformer-based LLM.
+    This works for text-only generation and image-text generation.
+
+    This *does not* currently support the following features:
+        - torch.compile
+        - quantization through torchao
+        - multi-GPU generation
```

Comment:
should update this and call out that this can be run distirbuted for larger models using TP. And could point to some pytorch docs on TP